pydicom/tests/test_filewriter.py

```diff
@@ -222,6 +223,23 @@ def test_write_empty_sequence(self):
         ds = read_file(self.file_out)
         assert ds.PerformedProcedureCodeSequence == []
 
+    def test_write_deflated(self):
+        """Read a Deflated Explicit VR Little Endian file, write it,
+           and then read the output, to verify that the written file
+           is correct.
+           """
+        original = read_file(deflate_name)
+        original.save_as(self.file_out)
+
+        self.file_out.seek(0)
+        rewritten = read_file(self.file_out)
+
+        assert (original.file_meta.TransferSyntaxUID ==
+                rewritten.file_meta.TransferSyntaxUID)
+        assert len(original) == len(rewritten)
```

Comment:
Can you do `assert original == rewritten`? That would compare the whole dataset.

Reply:
I cannot. (I actually tried it at first.) The problem is that the pytest `assert` custoimization sees that `original` and `rewritten` are both `dict`s and tries to turn them into `set`s for comparison. Since `DataElement`s aren't hashable, this goes very poorly:

> E       AssertionError: assert (0008, 0016) ...62144 elements == (0008, 0016) ...62144 elements
> E         (pytest_assertion plugin: representation of details failed: d:\sandbox\pydicom\venv\lib\site-packages\_pytest\assertion\util.py:371: TypeError: unhashable type: 'DataElement'.
> E          Probably an object has a faulty __repr__.)

I'm happy to do something else. Options that I can think of:

1. ```python
     was_ written_the_same = original == rewritten
     assert was_written_the_same
    ```
    This will give the effect that I think you were going for, but the feedback on a failure is abysmal. You know that the results weren't the same, and that's it
2. provide a more sophisticated matcher. I was thinking that if we compare `list(original.file_meta) + list(orginal)` (assuming file_meta exists, which I'd check), you get a better reporting experience - missing elements are noted, as are differing ones. It's not perfect, as I expect sequences containing `DataSet`s will have the original problem. I'd be happy to pursue a recursive converter, if you like this approach (or just leave it as is). I think this route is probably the best approach if we're interested in comparing `DataSet`s for equality long-term
3. compromise: instead of `list`ing the objects, `str` them. It will give an approximation, but will not be good at determining differences in long arrays of OBs or OWs, such as the PixelData

I'm happy to go with any of the above or an alternative approach that you like. I will likely sketch out the listy approach out of curiosity, but if you hate it, do stop me.

Reply:
Ah. And actually a straight `==` doesn't work on the two `FileDataset`s because the `.filename`s are different (this took a while to track down). This is workaroundable, either by just removing the `filename`, because we don't care in this case, which feels _a little_ dirty, but maybe is fine. I'll think more about it tomorrow.

Reply:
This was just a suggestion - we do this comparison of whole datasets in several tests. I hadn't thought about the file dataset issue - I will be fine with both removing the filename, or just comparing a few items.
What I usually do while developing, and the tests not passing yet - I iterate over all tags in both datasets and compare them:
```python
    for e1, e2 in zip(ds1, ds2):
        assert e1 == e2
```
 
After the issue has been fixed, I simplify that.

Reply:
I'd like a test to confirm that the dataset actually has been written deflated, even something simple like moving to the byte offset of the start of the deflated data then asserting that running `zlib.decompress` doesn't raise an exception.

Reply:
Also, I thought we fixed it so that dataset comparison only worked on elements?

Reply:
> I thought we fixed it so that dataset comparison only worked on elements

I don't think so. The comparison does fail until I remove the filenames. 
`FileDataset.__eq__` runs `_dict_equal` on `__dict`, excluding only `_dict`, so it would check `filename`.

Reply:
>  a test to confirm that the dataset actually has been written deflated

Added.

Reply:
> Maybe you can use this to compare the file datasets.

@mrbean-bremen, I'd already gone another way before reading your e-mail. My method to make a safe version of the DS lets pytest give pretty good info about elements that are new, missing, or changed, but if you don't like anything about it, I can for sure switch to drop the filename and just use `==` and assert against the bool, or whatever you like.